File: apps/web/src/lib/api/dev.ts
import { api } from './client';

export interface DevInfoResponse {
  image_hash: string;
  image_digest: string | null;
  image_tag: string;
  build_time: string | null;
  git_commit: string;
  git_branch: string;
  python_version: string;
  fastapi_version: string;
  db_migration_head: string | null;
  db_pool_size: number | null;
  db_checked_out: number | null;
  redis_connected: boolean;
  minio_connected: boolean;
  health_ready: boolean;
  uptime_seconds: number;
}

export interface WorkerRunResponse {
  id: string;
  worker_name: string;
  job_kind: string;
  trigger_source: string;
  status: string;
  started_at: string;
  finished_at: string | null;
  scope_user_id: string | null;
  result: Record<string, unknown>;
  error_message: string | null;
}

export interface WorkerRunsListResponse {
  items: WorkerRunResponse[];
  ops_ready: boolean;
}

export interface WorkerRunsLatestResponse {
  daily_bundle: WorkerRunResponse | null;
  fleet_insights: WorkerRunResponse | null;
  user_insights: WorkerRunResponse | null;
  ops_ready: boolean;
}

export interface DevInsightsRunResponse {
  status: 'ok';
  eligible_users: number;
  processed_users: number;
  failed_users: number;
  generated_insights: number;
}

export interface DevDbBackupItem {
  name: string;
  size_bytes: number;
  created_at: string;
  meta: Record<string, unknown> | null;
}

export interface DevDbBackupListResponse {
  items: DevDbBackupItem[];
  backup_dir: string;
  ops_ready: boolean;
  encryption_key_required: boolean;
}

export interface DevDbBackupCreateResponse {
  status: 'ok';
  backup: DevDbBackupItem;
  message: string;
}

export interface DevDbRestoreResponse {
  status: 'ok';
  restored: string;
  message: string;
  ops_ready: boolean;
}

export async function fetchDevInfo(signal?: AbortSignal): Promise<DevInfoResponse> {
  return api.get<DevInfoResponse>('/dev/info', { signal });
}

export async function fetchWorkerRuns(options?: {
  limit?: number;
  scope?: 'all' | 'me';
  signal?: AbortSignal;
}): Promise<WorkerRunsListResponse> {
  const params = new URLSearchParams();
  if (options?.limit != null) params.set('limit', String(options.limit));
  if (options?.scope) params.set('scope', options.scope);
  const query = params.toString();
  return api.get<WorkerRunsListResponse>(`/dev/workers${query ? `?${query}` : ''}`, {
    signal: options?.signal,
  });
}

export async function fetchWorkerRunsLatest(
  signal?: AbortSignal
): Promise<WorkerRunsLatestResponse> {
  return api.get<WorkerRunsLatestResponse>('/dev/workers/latest', { signal });
}

export async function runDevInsightsOnce(signal?: AbortSignal): Promise<DevInsightsRunResponse> {
  return api.post<DevInsightsRunResponse>('/dev/workers/insights/run-once', undefined, { signal });
}

export async function fetchDevDbBackups(signal?: AbortSignal): Promise<DevDbBackupListResponse> {
  return api.get<DevDbBackupListResponse>('/dev/db/backups', { signal });
}

export async function createDevDbBackup(signal?: AbortSignal): Promise<DevDbBackupCreateResponse> {
  return api.post<DevDbBackupCreateResponse>('/dev/db/backups', undefined, { signal });
}

export async function restoreDevDbBackup(
  name: string,
  signal?: AbortSignal
): Promise<DevDbRestoreResponse> {
  return api.post<DevDbRestoreResponse>('/dev/db/restore', { name, confirm: true }, { signal });
}

File: backend/app/schemas/dev.py
"""Developer diagnostics schemas.

The developer view is feature-flagged and authenticated because it exposes
deployment and infrastructure metadata. Worker/DB tooling stays under the same
gate; backup/restore additionally requires APP_ENV=development.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

DevHealthStatus = Literal["ok", "degraded", "down", "unavailable"]


class DevHealthComponent(BaseModel):
    """Reachability of one stack service as seen from the API process.

    This is an application-level probe (TCP / SELECT 1 / PING), not Docker
    HEALTHCHECK state. ADR-0015 forbids a Docker socket in the API container.
    """

    name: str
    status: DevHealthStatus
    detail: str = ""


class DevContainerHealth(BaseModel):
    """Docker-reported state for one Compose/stack container.

    ``issue`` is ``none`` for a clean one-shot (migrate exited 0) even though
    the container is stopped.
    """

    name: str
    service: str
    state: str
    health: Literal["healthy", "unhealthy", "starting", "none"] = "none"
    exit_code: int | None = None
    issue: Literal["unhealthy", "stopped", "none"] = "none"
    status_text: str = ""


class DevInfoResponse(BaseModel):
    image_hash: str
    image_digest: str | None
    image_tag: str
    build_time: str | None
    git_commit: str
    git_branch: str
    python_version: str
    fastapi_version: str
    db_migration_head: str | None
    db_pool_size: int | None
    db_checked_out: int | None
    redis_connected: bool
    minio_connected: bool
    health_ready: bool
    uptime_seconds: int
    health_components: list[DevHealthComponent] = Field(default_factory=list)
    containers: list[DevContainerHealth] = Field(default_factory=list)


class WorkerRunResponse(BaseModel):
    id: UUID
    worker_name: str
    job_kind: str
    trigger_source: str
    status: str
    started_at: datetime
    finished_at: datetime | None
    scope_user_id: UUID | None
    result: dict[str, Any]
    error_message: str | None


class WorkerRunsListResponse(BaseModel):
    items: list[WorkerRunResponse]
    ops_ready: bool = Field(
        default=False,
        description="Reserved for future production ops tooling; always false in v1.",
    )


class WorkerRunsLatestResponse(BaseModel):
    daily_bundle: WorkerRunResponse | None
    fleet_insights: WorkerRunResponse | None
    user_insights: WorkerRunResponse | None
    ops_ready: bool = False


class DevInsightsRunResponse(BaseModel):
    status: Literal["ok"] = "ok"
    eligible_users: int
    processed_users: int
    failed_users: int
    generated_insights: int


class DevDbBackupItem(BaseModel):
    name: str
    size_bytes: int
    created_at: datetime
    meta: dict[str, Any] | None = None


class DevDbBackupListResponse(BaseModel):
    items: list[DevDbBackupItem]
    backup_dir: str
    ops_ready: bool = False
    encryption_key_required: bool = True


class DevDbBackupCreateResponse(BaseModel):
    status: Literal["ok"] = "ok"
    backup: DevDbBackupItem
    message: str


class DevDbRestoreRequest(BaseModel):
    name: str
    confirm: bool = False


class DevDbRestoreResponse(BaseModel):
    status: Literal["ok"] = "ok"
    restored: str
    message: str
    ops_ready: bool = False
